src/cluster_undersample/cluster_undersample.jl

```diff
@@ -0,0 +1,207 @@
+"""
+Assuming that all the observations in the observation matrix X belong to the same class, undersample
+the observations using clustering undersampling
+
+# Arguments
+- `X`: A matrix where each column is an observation of floats
+- `n`: Number of observations in the undersampled dataset
+- `mode::AbstractString="nearest`: If `center` then the undersampled data will consist of the centriods of 
+    each cluster found. Meanwhile, if `nearest` then it will consist of the nearest neighbor of each centroid.
+- `maxiter::Integer=100`: Maximum number of iterations to run K-means
+- `rng::Integer=42`: Random number generator seed
+# Returns
+- `Xnew`: A matrix containing the undersampled observations
+"""
+function cluster_undersample_per_class(
+    X::AbstractMatrix{<:Real},
+    n::Integer;
+    mode::AbstractString = "nearest",
+    maxiter::Integer = 100,
+    rng::Integer = 42,
+)
+    (mode in ["center", "nearest"]) ||
+        throw(ArgumentError("mode must be either 'center' or 'nearest'"))
+    # to undersample down to n points find k=n clusters
+    seed!(rng)               # kmeans offers no better way :(
+    result = kmeans(X, n; maxiter)
+    # the n cluster centers are the undersampled points
+    X_new = result.centers
+    # unless mode is "nearest" where we seek the nearest neighbor of each center
+    if mode == "nearest"
+        tree = BallTree(X)
+        keep_inds, _ = knn(tree, X_new, 1, true)
+        keep_inds = vcat(keep_inds...)      # flatten
+        X_new = X[:, keep_inds]
+    end
+    return X_new
+end
+
+"""
+    cluster_undersample(
+        X, y; 
+        mode= "nearest", ratios = 1.0, maxiter = 100,
+        rng=default_rng(), try_preserve_type=true
+    )
+
+
+# Description
+
+Undersample a dataset using clustering undersampling as presented in [1] using K-means as the clustering algorithm.
+
+# Positional Arguments
+
+$(COMMON_DOCS["INPUTS"])
+
+# Keyword Arguments
+
+- `mode::AbstractString="nearest`: If `center` then the undersampled data will consist of the centriods of 
+    each cluster found. Meanwhile, if `nearest` then it will consist of the nearest neighbor of each centroid.
```

Comment:
```suggestion
- `mode::AbstractString="nearest`: If `"center"` then the undersampled data will consist of the centriods of 
    each cluster found. Meanwhile, if `"nearest"` then it will consist of the nearest neighbor of each centroid.
```